src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp

```diff
@@ -205,10 +212,28 @@ bool MergeTreeIndexConditionAnnoy::mayBeTrueOnGranule(MergeTreeIndexGranulePtr /
 
 bool MergeTreeIndexConditionAnnoy::alwaysUnknownOrTrue() const
 {
-    return condition.alwaysUnknownOrTrue("L2Distance");
+    return condition.alwaysUnknownOrTrue(metric_name);
 }
 
 std::vector<size_t> MergeTreeIndexConditionAnnoy::getUsefulRanges(MergeTreeIndexGranulePtr idx_granule) const
+{
+    if (metric_name == "L2Distance")
+    {
+        return getUsefulRangesImpl<::Annoy::Euclidean>(idx_granule);
+    }
+    else if (metric_name == "cosineDistance")
+    {
+        return getUsefulRangesImpl<::Annoy::Angular>(idx_granule);
+    }
+    else
+    {
+        throw Exception(ErrorCodes::INCORRECT_QUERY, "Wrong metric type {}", metric_name);
```

Comment:
`ErrorCodes::BAD_ARGUMENTS` would be more specific.

`Wrong metric type` --> `Unknown metric argument, must be 'L2Distance' or 'cosineDistance'` (slightly friendlier)

Also, the docs talk about `distance` ... suggest to keep this terminology.

(same for `throw` statements below)